From 8a928d9df229e47ad4c28d10e8c9696f06b8e2bd Mon Sep 17 00:00:00 2001
From: Joerg Hofrichter <joerg.hofrichter@ni.com>
Date: Tue, 28 May 2024 15:58:38 +0200
Subject: [PATCH] mpm: deactivate Duplicate Address Detection (DAD) on int0
 interface

Starting from systemd v245, Duplicate Address Detection (DAD) is automatically
enabled for static IPv4 addresses if the IP address is a link local address.
DuplicateAddressDetection=none (or DuplicateAddressDetection=true for historical
reasons) deactivates this.

Important notes on compatibility:
- systemd v243.2 (Yocto release zeus) does not support DuplicateAddressDetection=none
  and errors out without applying the IPv4 address -> use the historical
  parameter DuplicateAddressDetection=true to remain backward compatibility
- systemd v250.5 (Yocto release kirkstone) ignores the DuplicateAddressDetection=none
  (or DuplicateAddressDetection=true) parameter for link local addresses and enables
  DAD unconditionally. This is fixed by commit a9efd16090baee492e9dd9ae832b5b548e7c6fcf
  ("network: do not enable IPv4 ACD for IPv4 link-local address if ACD is disabled explicitly")
  which is not part of default Yocto systemd v250.5. In meta-ettus we have applied
  the patch so that the DuplicateAddressDetection=none (or DuplicateAddressDetection=true)
  parameter is honored for link local addresses
---
 mpm/systemd/network/int0.network | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/mpm/systemd/network/int0.network b/mpm/systemd/network/int0.network
index 67a04c322..79e352b2b 100644
--- a/mpm/systemd/network/int0.network
+++ b/mpm/systemd/network/int0.network
@@ -2,8 +2,13 @@
 Name=int0
 
 [Network]
-Address=169.254.0.1/24
 IPForward=ipv4
 
+[Address]
+Address=169.254.0.1/24
+# The setting actually means DuplicateAddressDetection=none. We use the old flag here
+# on purpose to remain backward compatibility with systemd v243.2 (Yocto release zeus)
+DuplicateAddressDetection=true
+
 [Link]
 MTUBytes=9000
-- 
2.34.1

